h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -160,6 +173,124 @@ public long getModelCount() {
+      // Attempt to train next model
+      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+      if (nextModelParams != null && isThereEnoughTime() && !_job.stop_requested()) {
+        parallelModelBuilder.run(ModelBuilder.make(nextModelParams));
+      } else {
+        parallelModelBuilder.noMoreModels();
+      }
+    }
+
+    private void constructScoringInfo(final Model model) {
+      ScoringInfo scoringInfo = new ScoringInfo();
+      scoringInfo.time_stamp_ms = System.currentTimeMillis();
+
+      model.fillScoringInfo(scoringInfo);
+      grid.setScoringInfos(ScoringInfo.prependScoringInfo(scoringInfo, grid.getScoringInfos()));
+      ScoringInfo.sort(grid.getScoringInfos(), _hyperSpaceWalker.search_criteria().stopping_metric());
+    }
+
+    private boolean isThereEnoughTime() {
+      final boolean enoughTime = hyperspaceIterator.max_runtime_secs() > 0 && hyperspaceIterator.time_remaining_secs() > 0;
+
+      if (!enoughTime) {
+        Log.info("Grid max_runtime_secs of " + hyperspaceIterator.max_runtime_secs() + " secs has expired; stopping early.");
+      }
+
+      return enoughTime;
+    }
+
+    private MP getNextModelParams(final HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceWalker, final Model model, final Grid grid){
+      MP params = null;
+
+      while (params == null) {
+        if (hyperSpaceWalker.hasNext(Optional.of(model))) {
+          params = hyperSpaceWalker.nextModelParameters(Optional.ofNullable(model));
+          final Key modelKey = grid.getModelKey(params.checksum(IGNORED_FIELDS_PARAM_HASH));
+          if(modelKey != null){
+            params = null;
+          }
+        } else {
+          break;
+        }
+      }
+
+      return params;
+    }
+  }
+
+  /**
+   * Searches the hyperspace and builds models in a parallel way - building the models in parallel.
+   *
+   * @param grid Grid to add models to
+   */
+  private void parallelGridSearch(final Grid<MP> grid) {
+    final HyperSpaceWalker.HyperSpaceIterator<MP> iterator = _hyperSpaceWalker.iterator();
+    final ModelFeeder modelFeeder = new ModelFeeder(iterator, grid);
+    final ParallelModelBuilder parallelModelBuilder = new ParallelModelBuilder(modelFeeder::onModelSuccessfullyBuilt,
+            modelFeeder::onModelBuildFailed);
+
+    List<ModelBuilder> startModels = new ArrayList<>();
+    final int parallelismLevel = _parallelismLevel > 1 ? _parallelismLevel : 2 * H2O.NUMCPUS;
+    final List<MP> mps = iterator.initialModelParameters(parallelismLevel);
+
+    for (int i = 0; i < mps.size(); i++) {
+      final MP nextModelParameters = mps.get(i);
+      final long checksum = nextModelParameters.checksum(IGNORED_FIELDS_PARAM_HASH);
+      if (grid.getModelKey(checksum) == null) {
+        startModels.add(ModelBuilder.make(nextModelParameters));
+      } else {
+        //TODO: Compensate for this scenario - we need to keep the level of paralelism at pre-set level
+      }
+    }
```

Comment:
not a big fan of this approach.
I think you could use an ArrayBlockingQueue limited to parallelism, and fill it continuously as soon as it has capacity (waiting when it's full) and as soon as you have next model to train.

Reply:
This would also solve your TODO

Reply:
another advantage of queues here: you don't need your callbacks to "know" what to do next:
like in `onModelBuildFailed`:
```      attemptBuildNextModel(parallelModelBuilder, null);```

this simplifies the organization of the code: communication between components is unidirectional, producer feeds the queue, consumer consumes, and possible a last queue consumes all the callbacks. No locks, no dialogue between components...

Reply:
I love queues, but it's really because they simplify things a lot most of the time.

Reply:
@pavel
> I wanted to get rid of any threads waiting in loop to do something...

But there's no interruption every `t` millis with queues. It's event-driven, the event is just encapsulated by the queue itself... When the queue is full, the producer just sleeps until it get notified that it can add more. so that producer thread wouldn't be "busy-waiting". The consumer on the other side, could be the main/current thread without any problem.

> Of course, a callback doesn't need to know what to do next, but these callbacks are intended to know what to do next. They're free to invoke other builders

It's just a question of separation of concerns. Your current callbacks manage 2 concerns: (1) adding the new model to the grid, (2) trigger the build for the next model (this concern being also partly managed by your initialization code).
Queue clearly separate those concerns as it would take care fully and only of providing model builders.
Of course, there's always some logic needed to decide when to stop: and in this case, this should be the consumer=current thread=thread responsible to start the training jobs. It takes a ModelBuilder from the queue, checks it (for example, if queue was fed with `Optional.empty()`, then it stops everything) as well as the time/max models conditions. If conditions met, it starts training, otherwise, it can interrupt the producer thread...

> Current version, despite considered as "ugly" by some, gives the callback the will to spawn n new jobs at a time.

This would clearly break the SoC principle imo, I'd not recommend this. I think there should always be one single place deciding what to do next. 

> Thinking of it as a queue won't allow me to get rid of any business logic and locks. 

Correct, the feeding queue won't allow you to get rid of the locks that you need to fill the grid. Maybe my sentence was not clear... was just saying that callbacks themselves could be queued, in which case you don't need the locks.
But I don't see this last queue as interesting as the feeding one: outside getting rid of the locks, it doesn't bring much. The feeding queue, on the other side has more benefits.

> Or, the queue can be serviced completely thread-free by callbacks only

You can't do that with blocking queues.

Reply:
After reading through it, I still see no point a (fixed-sized) queue. All I need is `n` models training in parallel. And the `n` can change. Queue has no benefits over this model to me, other than subjective benefits of "better readability" and other stuff like that.

Concerns are separable here as well - just introduce another method to do it.

My scenario: Model finishes building -> model is saved -> `n` new models are started, potentially based on what has just been finished building.
Queue scenario: Model is finished building -> waiting thread is notified -> model or `n` new models are started, but taken from the queue, no direct (time-wise) feedback ( (model length, performance of the latest model ...).

Not saying queues are a bad thing, I just don't see they're somehow superior. If you're ok with it, I'd leave this part as-is.